cpp/test/neighbors/ann_cagra_vpq.cuh

```diff
@@ -0,0 +1,337 @@
+                                  raft::make_const_mdspan(host_queries.view()),
+                                  raft::make_const_mdspan(host_index_candidate.view()),
+                                  host_indices_Cagra_view,
+                                  host_dists_Cagra_view,
+                                  ps.metric);
+
+          raft::copy(indices_dev.data(),
+                     host_indices_Cagra_view.data_handle(),
+                     ps.k * ps.n_queries,
+                     stream_);
+          raft::copy(distances_dev.data(),
+                     host_dists_Cagra_view.data_handle(),
+                     ps.k * ps.n_queries,
+                     stream_);
+          resource::sync_stream(handle_);
+        }
+      }
+
+      double min_recall = ps.min_recall;
+      EXPECT_TRUE(eval_neighbours(indices_naive,
+                                  indices_Cagra,
+                                  distances_naive,
+                                  distances_Cagra,
+                                  ps.n_queries,
+                                  ps.k,
+                                  0.003,
+                                  min_recall));
+      EXPECT_TRUE(eval_distances(handle_,
+                                 database.data(),
+                                 search_queries.data(),
+                                 indices_dev.data(),
+                                 distances_dev.data(),
+                                 ps.n_rows,
+                                 ps.dim,
+                                 ps.n_queries,
+                                 ps.k,
+                                 ps.metric,
+                                 1.0e-4));
+    }
+  }
+
+  void SetUp() override
+  {
+    database.resize(((size_t)ps.n_rows) * ps.dim, stream_);
+    search_queries.resize(ps.n_queries * ps.dim, stream_);
+    GenerateDataset(database.data(),
+                    search_queries.data(),
+                    ps.n_rows,
+                    ps.n_queries,
+                    ps.dim,
+                    static_cast<std::size_t>(std::sqrt(ps.n_rows)),
+                    stream_);
+    resource::sync_stream(handle_);
+  }
+
+  void TearDown() override
+  {
+    resource::sync_stream(handle_);
+    database.resize(0, stream_);
+    search_queries.resize(0, stream_);
+  }
+
+ private:
+  raft::resources handle_;
+  rmm::cuda_stream_view stream_;
+  AnnCagraVpqInputs ps;
+  rmm::device_uvector<DataT> database;
+  rmm::device_uvector<DataT> search_queries;
+};
+
+const std::vector<AnnCagraVpqInputs> vpq_inputs = raft::util::itertools::product<AnnCagraVpqInputs>(
+  {100},                                              // n_queries
+  {1000, 10000},                                      // n_rows
+  {128, 256},                                         // dim
+  {8, 12},                                            // k
+  {2, 4},                                             // pq_len
```

Comment:
We should extend test coverage with different values of n_queries, n_rows, dim and k. Note that it should not be a tensor product. We can vary individual parameters one at a time

Reply:
At would also make sense to add refine_ratio as a test option here.